### DISCUSSION TITLE
Using Node.js built-in modules

### DISCUSSION THREAD

**Discussion by nshiab** [denoland/deno#25948] [General]
Let's say we have something like this in a script: `import os from "node:os";`

Does it mean that Deno imports the necessary functions or classes from Node.js? If so, does it mean Node.js is required to run the script?

Is there a better way? If we use just Deno functions and methods to do the same things as Node.js, how can we ensure that the module or library runs with Node.js, too?

I am considering using Deno to build and maintain my libraries. The v2 is really great. But I want to make sure that:
- Deno users won't need to install or worry about Node.js
- Node.js users will still be able to use the libraries

Thank you!

**Comment by marvinhagemeister** [comment]
When you import a `node:*` module in Deno, then you import code from Deno that mirrors the Node APIs. It doesn't shell out to Node itself. Therefore Node is not required when using Deno.

If you want to write a library that works in both runtimes, only going with `node:*` imports is one way to do that.

**Comment by nshiab** [comment]
Thank you for the explanation, @marvinhagemeister !

From what I understand, I could also use the Deno std modules. Node would be able to run them.

Is there a big difference between the two approaches?

**Reply by bartlomieju** [reply]
@nshiab only a few `deno_std` modules can currently run in Node.js, @kt3k can talk more about it

**Reply by nshiab** [reply]
Ah! You're right. I double-checked the [JSR page](https://jsr.io/@std). Many of them are compatible, but some, like [@std/fs](https://jsr.io/@std/fs), work just with Deno.

Thanks for the answer!

<img width="1182" alt="Screenshot 2024-09-30 at 7 37 51 PM" src="https://github.com/user-attachments/assets/6d9f55f6-99cd-4a0c-bc5a-533537566a88">
